profiling/src/collections/string_table/mod.rs

```diff
@@ -0,0 +1,279 @@
+        }
+    }
+}
+
+impl LendingIterator for StringTableIter {
+    type Item<'a> = &'a str where Self: 'a;
+
+    fn next(&mut self) -> Option<Self::Item<'_>> {
+        self.iter.next()
+    }
+
+    fn count(self) -> usize {
+        self.iter.count()
+    }
+}
+
+impl IntoLendingIterator for StringTable {
+    type Iter = StringTableIter;
+
+    fn into_lending_iter(self) -> Self::Iter {
+        StringTableIter::new(self)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn test_basics() {
+        let mut table = StringTable::new();
+        // The empty string should already be present.
+        assert_eq!(1, table.len());
+        assert_eq!(StringId::ZERO, table.intern(""));
+
+        // Intern a string literal to ensure ?Sized works.
+        let string = table.intern("datadog");
+        assert_eq!(StringId::from_offset(1), string);
+        assert_eq!(2, table.len());
+    }
+
+    #[track_caller]
+    fn test_from_src(src: &[&str]) {
+        // Insert all the strings.
+        let mut table = StringTable::new();
+        let n_strings = src.len();
+        for string in src {
+            table.intern(string);
+        }
+        assert_eq!(n_strings, table.len());
+
+        // Re-inserting doesn't change the size.
+        for string in src {
+            table.intern(string);
+        }
+        assert_eq!(n_strings, table.len());
+
+        // Check that they are ordered correctly when iterating.
+        let mut actual_iter = table.into_lending_iter();
+        let mut expected_iter = src.iter();
+        while let (Some(expected), Some(actual)) = (expected_iter.next(), actual_iter.next()) {
+            assert_eq!(*expected, actual);
+        }
+
+        // The iterators should be exhausted at this point.
+        assert_eq!(None, expected_iter.next());
+        assert_eq!(0, actual_iter.count());
+    }
+
+    #[test]
+    fn test_small_set_of_strings() {
+        let cases: &[_] = &[
+            "",
+            "local root span id",
+            "span id",
+            "trace endpoint",
+            "samples",
+            "count",
+            "wall-time",
+            "nanoseconds",
+            "cpu-time",
+            "<?php",
+            "/srv/demo/public/index.php",
+            "pid",
+            "/var/www/public/index.php",
+            "main",
+            "thread id",
+            "A\\Very\\Long\\Php\\Namespace\\Class::method",
+            "/",
+        ];
+        test_from_src(cases);
+    }
+
+    /// Test inserting strings from a real WordPress profile, although it's
+    /// unknown at this point which profiler version generated it. It's a small
+    /// sample. Here we're checking that we don't panic or otherwise fail, and
+    /// that the total number of strings and the bytes of those strings match
+    /// the profile.
+    #[test]
+    fn test_wordpress() {
+        test_from_src(&wordpress_test_data::WORDPRESS_STRINGS);
```

Comment:
We're only adding each string once, in order.  Would it make sense to randomly add strings multiple times too?

Reply:
The `test_from_src` function adds every string twice.